usr/lib/common/pin_get.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2022
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#define _POSIX_C_SOURCE 200809L
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/mman.h>
+
+#include <pin_get.h>
+
+struct pinbuf pinbuf_empty = {
+    .p = NULL,
+    .len = 0,
+};
+
+static void echo(bool on)
+{
+    struct termios term;
+
+    tcgetattr(fileno(stdin), &term);
+
+    if (on)
+        term.c_lflag |= ECHO;
+    else
+        term.c_lflag &= ~ECHO;
+
+    tcsetattr(fileno(stdin), TCSAFLUSH, &term);
+}
+
+const char *pin_prompt(struct pinbuf *buf, const char *msg)
+{
+    ssize_t n;
+
+    if (!buf)
+        return NULL;
+
+    printf("%s", msg);
+    fflush(stdout);
+    echo(false);
+
+    n = getline(&buf->p, &buf->len, stdin);
+
+    echo(true);
+    printf("\n");
+    fflush(stdout);
+
+    /* delayed getline() error handling */
+    if (n == -1)
+        return NULL;
+
+    /* strip CR/LF (if any) */
+    (buf->p)[strcspn(buf->p, "\r\n")] = '\0';
+
+    buf->len = strlen(buf->p);
+    return buf->p;
+}
+
+const char *pin_get_new(struct pinbuf *buf, const char *msg1, const char *msg2)
+{
+    const char *pin = NULL;
+    struct pinbuf buf2 = pinbuf_empty;
+
+    if (!pin_prompt(buf, msg1) ||
+        !pin_prompt(&buf2, msg2))
+        return NULL;
+
+    if ((buf->len && buf2.len) &&
+        (buf->len == buf2.len) &&
+        (!strncmp(buf->p, buf2.p, buf->len)))
+        pin = buf->p;
+
+    pinbuf_free(&buf2);
+    return pin;
+}
+
+static const char *pin_get(struct pinbuf *buf, const char *env, const char *msg)
+{
+    const char *pin;
+
+    pin = getenv(env);
+    if (pin)
+        return pin;
+
+    return pin_prompt(buf, msg);
+}
+
+const char *pin_get_user(struct pinbuf *buf)
+{
+    return pin_get(buf, "OCK_USER_PIN", "Enter the USER PIN: ");
+}
+
+const char *pin_get_so(struct pinbuf *buf)
+{
+    return pin_get(buf, "OCK_SO_PIN", "Enter the SO PIN: ");
+}
+
+void *pinbuf_alloc(struct pinbuf *buf, size_t len)
+{
+    void *rc;
+
+    if (!buf)
+        return NULL;
+
+    *buf = pinbuf_empty;
+
+    rc = calloc(sizeof(char), len);
+    if (!rc)
+        return NULL;
+
+    buf->p = rc;
+    buf->len = len;
+
+    return rc;
+}
+
+void pinbuf_free(struct pinbuf *buf)
+{
+    if (!buf)
+        return;
+
+    if (buf->p)
+        memset(buf->p, 0, buf->len);
```

Comment:
Use OPENSSL_cleanse here, memset may get optimized out by the compiler.

Reply:
I would prefer not to introduce a dependecy to libcrypto here, just for cleaning the memory. Yes, the current implementation can be removed by compiler optimizations. But if you look into the OPENSSL_cleanse implementation, it is relatively easy to prevent that. 

pin_get is meant to be used in all ock tools. Should they all have a depency to libcrypto, just to wipe out the pin buffer? I would perfer to prevent the compiler optimization for this single line instead.

Reply:
we already link libcrypto everywhere (as well as many other libraries...).
But if you find a way to hinder the compiler to not optimize that out, fine with me.